### ISSUE TITLE
bug: Qwen3-0.6B + --continuous-batching degrades after first request (sequential KV-cache corruption)

### ISSUE THREAD

**Issue by jackneil** [jackneil/vllm-mlx-patched#29]
## TL;DR

`vllm-mlx serve Qwen/Qwen3-0.6B-MLX-4bit --continuous-batching` degrades across sequential requests in the same process: **first request works, second+ requests produce token-repetition garbage** (Korean `작`, Chinese `笑话`, Cyrillic `н`, `ongoongo`, `ystone` etc). The same server without `--continuous-batching` handles 5+ sequential diverse prompts cleanly.

This is the same subsystem family as #380 ("continuous-batching yields nonsense output") and #384 ("Gemma 4 continuous batching leaks content across concurrent requests"). Different model (Qwen3-0.6B), different pattern (sequential degradation, not concurrent leak). Worth filing separately so it can be tracked.

## Environment

- **Chip:** Apple M3 Ultra (512 GB unified)
- **OS:** macOS 26.3 (build 25D125)
- **Python:** 3.12.13
- **mlx:** 0.31.1
- **mlx-lm:** 0.31.2
- **vllm-mlx:** 0.2.7 (editable install from `jackneil/vllm-mlx-patched@ec2169a`, rebased on `waybarrios/vllm-mlx@b4fa0308`)
- **Model:** `Qwen/Qwen3-0.6B-MLX-4bit` (official Qwen MLX 4-bit conversion, 0.3 GB on disk)

## Server launches (identical except for `--continuous-batching`)

**Broken (with batching):**
```
vllm-mlx serve Qwen/Qwen3-0.6B-MLX-4bit \
  --host 0.0.0.0 --port 8011 \
  --continuous-batching
```

**Working (without batching):**
```
vllm-mlx serve Qwen/Qwen3-0.6B-MLX-4bit \
  --host 0.0.0.0 --port 8011
```

(No reasoning-parser, no tool-call-parser, no thinking-budget flags. Bug reproduces on a completely bare launch.)

## Reproduction

Fire 3 sequential requests with different prompts to the same process.

```bash
for p in "What is 2+2?" "Say hello in three words." "Write a two-line Dr. Seuss rhyme about NASCAR."; do
  echo "=== $p ==="
  curl -s --max-time 30 -X POST http://localhost:8011/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d "{\"model\":\"Qwen/Qwen3-0.6B-MLX-4bit\",\"max_tokens\":80,\"temperature\":0.0,\"messages\":[{\"role\":\"user\",\"content\":\"$p\"}]}" \
    | python3 -c "import sys,json; d=json.load(sys.stdin); c=d['choices'][0]; print(repr(c['message']['content'][:200]))"
done
```

### Output — `--continuous-batching` ON

```
=== What is 2+2? ===
"<think>\nOkay, so I'm trying to figure out what 2+2 is. Let's start by breaking it down..."   ← COHERENT

=== Say hello in three words. ===
'ongoongoongoongoongoongoongoongoongoongoongoongoongoongoongo...'                              ← GARBAGE

=== Write a two-line Dr. Seuss rhyme about NASCAR. ===
'class MyClass {\n    public static void main(String[] args) {\n        // Create a new...'   ← OFF-TOPIC + degenerate
```

### Output — same 3 prompts, batching OFF (SimpleEngine)

```
=== What is 2+2? ===
"<think>\nOkay, so I need to figure out what 2+2 is. Let's start by thinking about what addition means..."  ← COHERENT

=== Say hello in three words. ===
'<think>\nOkay, the user wants me to say "say hello in three words." Let me think about...'              ← COHERENT

=== Write a two-line Dr. Seuss rhyme about NASCAR. ===
"<think>\nOkay, I need to write a two-line Dr. Seuss rhyme about NASCAR. Let's start by recalling..."     ← COHERENT
```

5 different prompts tested on the SimpleEngine path — all coherent. 3 sequential prompts tested on the batching path — request 1 coherent, 2+ degenerate.

## What was ruled out

- **Sampling / temperature.** Reproduces at `temperature=0.0` (greedy), `temperature=0.7`, `temperature=0.7 + top_p=0.9`, `temperature=0.7 + top_k=40`, and `temperature=0.7 + repetition_penalty=1.15`. Sweep at greedy shows the degeneration is deterministic on batching, not a sampling cliff.
- **Thinking / reasoning parser.** Bug reproduces with `--reasoning-parser qwen3` both ON and OFF. Also reproduces without the `--enable-auto-tool-choice` / `--tool-call-parser qwen3` flags.
- **`--max-thinking-token-budget`.** Bug reproduces with and without Layer 2 ceiling. Also with `/no_think` in the user message to bypass thinking entirely.
- **Quantization.** Reproduces on Qwen's official `Qwen/Qwen3-0.6B-MLX-4bit`. (Separately, `mlx-community/Qwen3-0.6B-8bit` is worse — same bug plus the model itself has calibration issues — but the 4-bit alone is sufficient to reproduce.)
- **Model-capacity floor.** 0.6B is a small model, but SimpleEngine path demonstrates it CAN produce coherent output across sequential requests. The failure is specifically introduced by the `--continuous-batching` engine.
- **Arena proxy / middleware.** Reproduces on direct curl to `:8011`, no arena in the path.
- **Cross-process state.** Fresh process per test. Bug reproduces on the SECOND request of a freshly-launched process (PID < 5s old).

## Hypothesis

The `--continuous-batching` engine's prefix-cache or KV-cache state is being written incorrectly after the first request when the attention heads / layer sizes of Qwen3-0.6B are small. Subsequent requests read the corrupted cache and produce garbage tokens. The fact that `prompt_tokens=18 cached=18 remaining=0` shows up in the scheduler logs on the second request (the full prompt prefix matched the cache from request 1) suggests the cache is believed to be valid by the scheduler even though the stored activations are corrupt.

Similar mechanism to #142/#136 (prefix cache tensor slicing on MoE 3D KV cache) but affecting 2D cache on tiny dense attention heads. The Qwen3-0.6B has **28 attention heads at 64 hidden dim per head** — unusually thin heads compared to typical models — which may interact badly with whatever tensor-copy path the batching engine uses.

## Known-good reference (SimpleEngine)

Drop `--continuous-batching` and the same process handles a stream of sequential requests cleanly. That's the current workaround used by `hank-llm-arena` for this specific model — the model card's "Batching" checkbox must be unchecked on launch.

## Context on impact

This isn't a low-impact edge case even though 0.6B is small:

1. Small Qwen3 variants (0.6B, 1.7B, 4B) are the natural targets for fine-tuning and speed benchmarking. Users who spin one up for quick evaluation via a typical `serve --continuous-batching` flow will see an inexplicably broken model after the second request.
2. Our arena `CLAUDE.md` rule says *"--continuous-batching on vllm-mlx is required for MORE than throughput: without it SimpleEngine runs (can't attach per-request logits processors), so thinking_token_budget silently no-ops."* Disabling batching to work around the bug costs us the thinking-budget enforcement path that Layers 1+2 of our Qwen3 runaway mitigation (`vllm-mlx-patched` #26 + #27) depend on.
3. Sequential degradation is harder to notice than concurrent corruption — an interactive user will just think the model is broken, because it is.

## Suggested investigation

1. Repro with `--log-level DEBUG` to capture the scheduler/prefix-cache events on requests 1, 2, 3. Compare the KV-cache tensor shapes before and after the first request completes.
2. Check if the prefix-cache store path handles small head-dim tensors correctly (Qwen3-0.6B: `num_attention_heads=16`, `num_key_value_heads=8`, `head_dim=128` per its config).
3. Run the same repro against other tiny Qwen3 models (1.7B, 4B) to see if the cliff is between 0.6B and 1.7B or affects all of them.
4. Compare `BatchedEngine.store_prefix_kv` against the MoE-fix that landed in #142 — may be an analogous bug on the dense path.

If you want the raw request/response traces I can paste them; didn't include to keep the issue tight.

**Comment by jackneil** [comment]
Fix in https://github.com/jackneil/vllm-mlx-patched/pull/30 — ports `waybarrios/vllm-mlx#385` + `waybarrios/vllm-mlx#365`, verified end-to-end against a live Qwen3-0.6B server.

**Comment by jackneil** [comment]
Verified fixed on arena (M3 Ultra, vllm-mlx-patched @ `1f333f1`, editable install).

Restarted `Qwen/Qwen3-0.6B-MLX-4bit` with `--continuous-batching` on port 8011 (bare `vllm-mlx serve` path, no arena middleware). Fired 7 sequential requests in the same process — 3 at `temperature=0.0`, 4 at `temperature=0.7` — covering radically different topics (math, colors, jokes, US history, Dr. Seuss parody).

**Before fix:** request 1 coherent, request 2 collapsed into `ongoongo…` / `笑话` / `н` / `ystone` repetition loops within ~5 tokens.
**After fix (#30):** all 7 requests produced coherent English thinking output. Request 7 reads like request 1 — no drift, no language flipping, no repetition loops.

Closing this. Happy to reopen if a regression surfaces during broader fleet testing.
